### PR TITLE
Split PRs and issues in the top search

### LINKED ISSUE #1635: Search results type for PR
Repository [search results](https://github.com/sindresorhus/refined-github/search?q=jerone) are currently categorized in 4 types: Code, Commits, Issues and Wikis (if available).
Where issues is a combination of issues and PR. My suggestion is that we make another type for all PR's and separate them out of the issues type.

### PR DESCRIPTION
**Pull request by jerone**
Repository search results are currently categorized in 4 types: Code, Commits, Issues and Wikis (if available).
Where issues is a combination of issues and PR. This PR will make another type for all PR's and separate them out of the issues type.

### Test
https://github.com/sindresorhus/refined-github/search?q=test+is%3Apr&type=Issues

### Screenshot
image

### Ref
* [x] Closes #1635

### CODE REVIEW

File: source/features/search-results-filter.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import {React} from 'dom-chef/react';
+import select from 'select-dom';
+import features from '../libs/features';
+import domify from '../libs/domify';
+
+function getSearchQuery() {
+	const url = new URL(location.href);
+	const query = url.searchParams.get('q');
+	return query;
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]:
```suggestion
	return new URLSearchParams(location.search).get('q');
```

File: source/features/search-results-filter.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+function createUrl(type) {
+	const query = getSearchQuery().replace(/\s?is:pr/gi, '').replace(/\s?is:issue/gi, '');
+	const url = `${location.origin}${location.pathname}?q=${encodeURIComponent(query)}+is%3A${type}&type=Issues`;
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: This should use proper URL construction using `URL()` and `URLSearchParams()`.

File: source/features/search-results-filter.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+async function fetchCount(type) {
+	const url = createUrl(type);
+	const response = await fetch(url, {credentials: 'same-origin'});
+	const text = await response.text();
+	const dom = domify(text);
+	const el = dom.querySelector('.menu a.selected .Counter');
+	const val = el.textContent;
+	const c = parseInt(val, 10);
+	return c;
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: Use a proper variable name.

File: source/features/search-results-filter.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+async function fetchCount(type) {
+	const url = createUrl(type);
+	const response = await fetch(url, {credentials: 'same-origin'});
+	const text = await response.text();
+	const dom = domify(text);
+	const el = dom.querySelector('.menu a.selected .Counter');
+	const val = el.textContent;
+	const c = parseInt(val, 10);
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: You can just pass `el.textContent` directly here.

File: source/features/search-results-filter.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	const url = `${location.origin}${location.pathname}?q=${encodeURIComponent(query)}+is%3A${type}&type=Issues`;
+	return url;
+}
+
+async function fetchCount(type) {
+	const url = createUrl(type);
+	const response = await fetch(url, {credentials: 'same-origin'});
+	const text = await response.text();
+	const dom = domify(text);
+	const el = dom.querySelector('.menu a.selected .Counter');
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: `el` => `element`

File: source/features/search-results-filter.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+function createUrl(type) {
+	const query = getSearchQuery().replace(/\s?is:pr/gi, '').replace(/\s?is:issue/gi, '');
+	const url = `${location.origin}${location.pathname}?q=${encodeURIComponent(query)}+is%3A${type}&type=Issues`;
+	return url;
+}
+
+async function fetchCount(type) {
+	const url = createUrl(type);
+	const response = await fetch(url, {credentials: 'same-origin'});
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: `{credentials: 'same-origin'}` is the default.

File: source/features/search-results-filter.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	const el = dom.querySelector('.menu a.selected .Counter');
+	const val = el.textContent;
+	const c = parseInt(val, 10);
+	return c;
+}
+
+async function init() {
+	const menu = select('nav.menu');
+	for (const link of select.all('a:not([href*="&type=Issues"])', menu)) {
+		link.href = link.href.replace(/\+?is%3Apr/gi, '').replace(/\+?is%3Aissue/gi, '');
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: `link.href` => `link.search`

Comment (sindresorhus) [reply source/features/search-results-filter.tsx]: And use `new URLSearchParams()`

File: source/features/search-results-filter.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	if (!issuesLink) { // Stop when issues are not enabled for this repo.
+		return;
+	}
+
+	issuesLink.href = createUrl('issue');
+
+	const prUrl = createUrl('pr');
+	const prHtml = <a class="menu-item" href={prUrl}>PR<span class="Counter ml-1 mt-1" data-search-type="PR"></span></a>;
+	issuesLink.after(prHtml);
+	if (getSearchQuery().includes('is:pr')) {
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: This is too naive. What if GitHub adds a `is:project` query?

File: source/features/search-results-filter.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		select('.selected', menu).classList.remove('selected');
+		prHtml.classList.add('selected');
+	}
+
+	select('.Counter', prHtml).textContent = await fetchCount('pr');
+	select('.Counter', issuesLink).textContent = await fetchCount('issue');
```
Comment (sindresorhus) [inline source/features/search-results-filter.tsx]: The counts should be fetched concurrently with `Promise.all()`.

**Comment by sindresorhus** [comment]
I think the menu title should be `Pull Requests`, not `PR`.

**Comment by sindresorhus** [comment]
On the PR tab, it says `7 issues in sindresorhus/refined-github`, but we should also change that to `7 pull requests in sindresorhus/refined-github`

**Comment by jerone** [comment]
@bfred-it & @sindresorhus I addressed all feedback; thank you. A lot is changed, so maybe another feedback round...

File: source/features/search-results-filter.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+	if (prIsActive()) {
+		select('.selected', menu).classList.remove('selected');
+		prHtml.classList.add('selected');
+
+		const h3 = select('.codesearch-results h3');
+		h3.innerHTML = h3.innerHTML.replace(/\bissues\b/gi, 'pull requests');
+	}
+
+	const [prCount = 0, issueCount = 0] = await Promise.all([fetchCount('pr'), fetchCount('issue')]);
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]: @sindresorhus I don't get TypeScript's requirement to have defaults in destructuring, especially if `fetchCount` _always_ returns a number.

Is it useful or it could it be disabled?

File: source/features/search-results-filter.tsx [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+	const prCounter = select('.Counter', prHtml);
+	if (prCounter) { // When there are no results, the counter is absent.
+		if (prCount > 0) {
+			prCounter.textContent = prCount;
+		} else {
+			prCounter.parentNode.removeChild(prCounter);
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]:
```suggestion
			prCounter.remove();
```

File: source/features/search-results-filter.tsx [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+		}
+	}
+
+	const issueCounter = select('.Counter', issueLink);
+	if (issueCounter) { // When there are no results, the counter is absent.
+		if (issueCount > 0) {
+			issueCounter.textContent = issueCount;
+		} else {
+			issueCounter.parentNode.removeChild(issueCounter);
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]:
```suggestion
			issueCounter.remove();
```

File: source/features/search-results-filter.tsx [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+async function fetchCount(type) {
+	const response = await fetch(createUrl(type));
+	const text = await response.text();
+	const counter = domify(text).querySelector('.menu a.selected .Counter');
+	return counter ? counter.textContent : 0; // When there are no results, the counter is absent.
+}
+
+function prIsActive() {
+	return getSearchQuery().split(' ').includes('is:pr');
```
[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/search-results-filter.tsx]: Or just `.split(/\s+/)`

[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/search-results-filter.tsx]: Or simply
```suggestion
	return /\bis:pr\b/.test(getSearchQuery());
```

File: source/features/search-results-filter.tsx [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+async function init() {
+	const menu = select('nav.menu');
+
+	const issueLink = select('a[href*="&type=Issues"]', menu);
+	if (!issueLink) { // Stop when issues are not enabled for this repo.
+		return;
+	}
+
+	const links = select.all('a', menu).filter(item => item !== issueLink);
+	cleanLinks(links);
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]: Since you're not using `links` anywhere else, you can move the `select` into `cleanLinks()`

File: source/features/search-results-filter.tsx [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+	issueLink.href = createUrl('issue');
+
+	const prUrl = createUrl('pr');
+	const prHtml = <a class="menu-item" href={prUrl}>Pull Requests<span class="Counter ml-1 mt-1" data-search-type="PR"></span></a>;
```
Comment (bfred-it) [inline source/features/search-results-filter.tsx]:
```suggestion
	const prHtml = (
		<a class="menu-item" href={prUrl}>
			Pull Requests
			<span class="Counter ml-1 mt-1" data-search-type="PR"/>
		</a>
	);
```

File: source/features/search-results-filter.tsx (hunk 11 above)
Comment (sindresorhus) [reply source/features/search-results-filter.tsx]: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/213

File: source/features/search-results-filter.tsx (hunk 10 above)
Comment (sindresorhus) [reply source/features/search-results-filter.tsx]: Probably because `fetchCount` currently returns `string | number`, see https://github.com/sindresorhus/refined-github/pull/1738#discussion_r249638241

File: source/features/search-results-filter.tsx [hunk 16]
```diff
@@ -0,0 +1,90 @@
+import {React} from 'dom-chef/react';
+import select from 'select-dom';
+import features from '../libs/features';
+import domify from '../libs/domify';
+
+function cleanLinks(links) {
+	for (const link of links) {
+		const href = new URL(link.href);
+		href.searchParams.set('q', cleanSearchQuery(href.searchParams.get('q')));
+		link.href = href;
```
[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/search-results-filter.tsx]: @sindresorhus not sure if you like it, but most URL props/classes accept objects as well:

`link.search = searchParams` and e.g. `new URL(location)` work perfectly — instead `link.search = searchParams.toString()` and `new URL(location.href)` respectively.

[the remark this answers is not in the document]
Comment (bfred-it) [reply source/features/search-results-filter.tsx]: Correction: `new URLSearchParams(link)` doesn't work correctly.

[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/features/search-results-filter.tsx]: > - link.search = searchParams.toString();
+ link.search = searchParams;

I'm aware that works, I just don't like the magic and prefer to be explicit about coercions.

File: source/features/search-results-filter.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (jerone) [reply source/features/search-results-filter.tsx]: Why would I loop through them to only check if the array contains an specific string...

File: source/features/search-results-filter.tsx [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+	const prUrl = createUrl('pr');
+	const prHtml = <a class="menu-item" href={prUrl}>Pull Requests<span class="Counter ml-1 mt-1" data-search-type="PR"></span></a>;
+	issueLink.after(prHtml);
+
+	if (prIsActive()) {
+		select('.selected', menu).classList.remove('selected');
+		prHtml.classList.add('selected');
+
+		const h3 = select('.codesearch-results h3');
+		h3.innerHTML = h3.innerHTML.replace(/\bissues\b/gi, 'pull requests');
```
[the remark this answers is not in the document]
Comment (jerone) [reply source/features/search-results-filter.tsx]: I stayed with the `innerHTML` solution, so I could easily implement the plural detection (`pull request` vs `pull requests`).
Let me know if you know a better solution to get the text node.

File: source/features/search-results-filter.tsx (hunk 10 above)
Comment (jerone) [reply source/features/search-results-filter.tsx]: Removed the defaults and made `fetchCount` always return a number.

File: source/features/search-results-filter.tsx (hunk 14 above)
Comment (jerone) [reply source/features/search-results-filter.tsx]: I need to filter out `issuesLink` from all links. What I did was remove the constant `links` and put the filter directly on the `cleanLinks` method.

**Comment by jerone** [comment]
@bfred-it & @sindresorhus I addressed all feedback again. Let me know if I need to change anything.

**Comment by bfred-it** [comment]
@jerone I made some changes (https://github.com/sindresorhus/refined-github/tree/pr/1738) but git won't let me push, did you disable the checkbox? (#1665)

**Comment by jerone** [comment]
[**@bfred-it**](https://github.com/bfred-it) commented on [Jan 23, 2019, 12:51 PM GMT+1](https://github.com/sindresorhus/refined-github/pull/1738#issuecomment-456773773 "2019-01-23T11:51:43Z - Replied by Github Reply Comments"):
> [@jerone](https://github.com/jerone) I made some changes ([<tt>91ee111</tt>](https://github.com/sindresorhus/refined-github/commit/91ee11157340e95dceb57abb447edaaaa06ac4af) ) but git won't let me push, did you disable the checkbox? ([#1665](https://github.com/sindresorhus/refined-github/issues/1665))

Yes, intentionally. I *learn so much more from the feedback* from the reviewers, then that someone completely rewrites the code that I contributed to this project. I understand that there are some code styling differences (open to any feedback), but there are multiple ways to get to end-goal (if it doesn't contain any bugs).

Anyways, I recognize your contributions and checked the "allow edits from maintainers" again. Feel free to push.

**Comment by jerone** [comment]
@bfred-it I can't get the latest code to work anymore...
image

File: source/features/search-results-filter.tsx [hunk 18]
```diff
@@ -1,95 +1,66 @@
...
+		.replace(/\bis:(pr|issue)\b/gi, newType)
+		.replace(/\s{2,}/g, ' ');
 }
 
-function createUrl(type) {
-	const url = new URL(location.pathname, location.origin); // Create url without any search params.
-	url.searchParams.set('q', cleanSearchQuery(getSearchQuery()) + ` is:${type}`);
+function createUrl(type, pathname = location.pathname) {
+	const url = new URL(pathname, location.origin);
+	url.searchParams.set('q', cleanSearchQuery(getSearchQuery(), `is:${type}`));
```
Comment (jerone) [inline source/features/search-results-filter.tsx]: This assumes that `is:type` is in the query, but that is not the case when freshly opening the search page.

File: source/features/search-results-filter.tsx [hunk 19]
```diff
@@ -1,95 +1,66 @@
...
-	const prHtml = (
-		<a class="menu-item" href={prUrl}>
-			Pull Requests
-			<span class="Counter ml-1 mt-1" data-search-type="PR" />
-		</a>
+	const prLink = issueLink.cloneNode(true);
+	prLink.textContent = 'Pull requests';
+	prLink.href = createUrl('pr');
+	prLink.append(
+		<include-fragment src={createUrl('pr', 'search/count')} />
```
Comment (jerone) [inline source/features/search-results-filter.tsx]: I seem to always get `8M` as a result, independent of my search query.

**Comment by jerone** [comment]
[**@jerone**](https://github.com/jerone) commented on [Jan 27, 2019, 6:48 PM GMT+1](https://github.com/sindresorhus/refined-github/pull/1738#issuecomment-457938723 "2019-01-27T17:48:24Z - Replied by Github Reply Comments"):
> @bfred-it I can't get the latest code to work anymore...


Fixed in commit https://github.com/sindresorhus/refined-github/pull/1738/commits/e356a74fc1a61dcb0d648e7c73f603195b8220c2

**Comment by bfred-it** [comment]
Good catch. I had only tested the global search

File: source/features/split-issue-pr-search-results.tsx [hunk 20]
```diff
@@ -54,6 +54,8 @@ function init() {
 
 		const title = select('.codesearch-results h3').firstChild;
 		title.textContent = title.textContent.replace('issue', 'pull request');
+	} else if (/\btype=Issues\b/.test(location.search) && !/\bis:issue\b/.test(getSearchQuery())) {
+		issueLink.click();
```
Comment (bfred-it) [inline source/features/split-issue-pr-search-results.tsx]: We can't do this, it will cause a refresh.

It's best to update the links on the page:

https://github.com/sindresorhus/refined-github/blob/329c2f25e9d85a62b0bc450d8e21507ab1d1a3e1/source/features/sort-milestones-by-closest-due-date.tsx

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: [**@bfred-it**](https://github.com/bfred-it) commented on [Jan 31, 2019, 6:14 AM GMT+1](https://github.com/sindresorhus/refined-github/pull/1738#discussion_r252537327 "2019-01-31T05:14:23Z - Replied by Github Reply Comments"):
> But first, how do you get to this point?
> 
> /\\btype=Issues\\b/.test(location.search) && !/\\bis:issue\\b/.test(getSearchQuery()))

Steps to reproduce:
1. Go to https://github.com/search?q=typescript&type=Issues
2. image

Any suggestions to correct the number in the results?

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Yes but who do you get to that URL? Can you only reach it by linking to it directly?

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: In that case _perhaps_ we shouldn't alter the results, because someone linked to both PRs and Issues at the same time.

To do this:

- de-select the Issues nav
- add ` and pull requests` to the title

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: The same can be reproduced by:
1. Go to https://github.com
2. Do a global search for `script`
3. Select "Issues" type
4. Alter the search value to `typescript`
5. Same results... image

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Uhm, that doesn't happen because we already fix the "Issues" link on that page

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Indeed. We should capture `submit` and alter the `q` to match the current page (issues or PR)

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Sounds like the logical choice, but we're actually not in either link: clicking either one of those will show different content.

This (https://github.com/sindresorhus/refined-github/pull/1738#discussion_r253299205) is not too difficult to do and works for both the cases we talked about. The `submit` capture would be a plus, but that situation is also covered by (https://github.com/sindresorhus/refined-github/pull/1738#discussion_r253299205)

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Ehhhh I just saw your screenshot. Maybe it doesn't look too bad. Sometimes the "current page" highlighted in a menu doesn't actually link to itself but to the "top" view. So I'd be ok with your solution.

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: [**@bfred-it**](https://github.com/bfred-it) commented on [Feb 3, 2019, 1:36 PM GMT+1](https://github.com/sindresorhus/refined-github/pull/1738#discussion_r253300337 "2019-02-03T12:36:08Z - Replied by Github Reply Comments"):
> Ehhhh I just saw your screenshot. Maybe it doesn't look too bad.
> 
> Sometimes the "current page" highlighted in a menu doesn't actually link to itself but to the "top" view. So I'd be ok with your solution.

I kinda like it too. It gives the user feedback that both results are shown (issues & PRs). But it also keeps the original urls, linking to their corresponding results (issues or PRs).

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: Committed the change for selecting both menu items when the user has not decided which type should be used (issues or PRs), without reloading the page.

**Comment by jerone** [comment]
I think we tackled every situation. 👍  :shipit:

File: source/features/split-issue-pr-search-results.tsx [hunk 21]
```diff
@@ -52,8 +52,10 @@ function init() {
 		issueLink.classList.remove('selected');
 		title.textContent = title.textContent.replace('issue', 'pull request');
 	} else if (/\btype=Issues\b/.test(location.search) && !/\bis:issue\b/.test(getSearchQuery())) {
+		title.textContent = title.textContent
+			.replace(/issue\b/, 'issue or pull request')
+			.replace(/issues/, 'issues and pull requests');
```
Comment (bfred-it) [inline source/features/split-issue-pr-search-results.tsx]: Test: https://github.com/search?q=alsdka&type=Issues

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: `'issues and/or pull requests'` ...

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: lol the link was supposed to only show one result, but pasting the link here created a second result. Check out now

Comment (jerone) [reply source/features/split-issue-pr-search-results.tsx]: Oh lol, I was actually suggesting an sentence change. That [previous link](https://github.com/search?q=alsdka&type=Issues) showed two results, both of them in pull requests. That's why I suggested to reword it to `'issues and/or pull requests'`, because there were zero (0) issues.

Comment (bfred-it) [reply source/features/split-issue-pr-search-results.tsx]: Nah I think it’s ok with “and” since the search contains both types. Also easier to read

**Comment by bfred-it** [comment]
a

**Comment by bfred-it** [comment]
Finally merged! Thanks @jerone 🌝
